### PR TITLE
Completes basic dtypes for all_reduce api in eager mode

### PR DESCRIPTION
**Pull request by HermitSun**
### PR types

New features

### PR changes

OPs

### Describe

This pr completes the basic function of communication framework, support various data types.
- Supports int8, uint8, bool dtypes for all_reduce in NCCL and GLOO backends.
- Completes to test all supported dtypes in `all_reduce` in eager mode.

通信框架功能进一步补全，通信操作支持传输丰富的数据类型。
- 对`all_reduce`操作，支持动态图场景下使用NCCL、GLOO后端进行int8、uint8、bool类型的数据传输。
- 补充动态图场景下已支持的所有数据类型的`all_reduce`的测试。
- 对应文档的中文api修改见https://github.com/PaddlePaddle/docs/pull/5206

### CODE REVIEW

File: python/paddle/distributed/collective.py [hunk 1]
```diff
@@ -785,8 +785,8 @@ def all_reduce(tensor, op=ReduceOp.SUM, group=None, use_calc_stream=True):
 
     Args:
         tensor (Tensor): The input Tensor. It also works as the output Tensor. Its data type
-            should be float16, float32, float64, int32 or int64.
-        op (ReduceOp.SUM|ReduceOp.MAX|ReduceOp.Min|ReduceOp.PROD): Optional. The operation used. Default value is ReduceOp.SUM.
+            should be float16, float32, float64, int32, int64, int8, uint8 or bool.
+        op (ReduceOp.SUM|ReduceOp.MAX|ReduceOp.MIN|ReduceOp.PROD): Optional. The operation used. Default value is ReduceOp.SUM.
```
Comment (LiYuRio) [inline python/paddle/distributed/collective.py:790]: 可以把下面例子里，那个古早的依赖numpy的写法修改了，现在to_tensor可以支持直接传入一个python list来构建tensor了

Comment (HermitSun) [reply python/paddle/distributed/collective.py:790]: 已修改

File: python/paddle/fluid/tests/unittests/test_collective_allreduce_api.py [hunk 2]
```diff
@@ -41,6 +41,98 @@ def test_allreduce_gloo(self):
         self.check_with_place("collective_allreduce_api.py", "allreduce",
                               "gloo", "2")
 
+    def test_allreduce_nccl_dygraph(self):
+        self.check_with_place("collective_allreduce_api_dygraph.py",
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_collective_allreduce_api.py]: One loop of an array is enough.

File: python/paddle/distributed/collective.py [hunk 3]
```diff
@@ -984,7 +981,7 @@ def all_gather(tensor_list, tensor, group=None, use_calc_stream=True):
     """
 
     Gather tensors from all participators and all get the result. As shown
-    below, 4 GPUs each start 4 processes and the data on each GPU is represnted
+    below, 4 GPUs each starts 4 processes and the data on each GPU is represented
```
Comment (gongweibao) [inline python/paddle/distributed/collective.py:983]: One process is started with a GPU and the data of this process is represented by the group rank.

**Review by gongweibao — CHANGES_REQUESTED** [review]
To add support of `bool` type, should we modify c++ backend?

File: python/paddle/fluid/tests/unittests/test_collective_allreduce_api.py (hunk 2 above)
Comment (HermitSun) [reply python/paddle/fluid/tests/unittests/test_collective_allreduce_api.py]: Modified.

File: python/paddle/distributed/collective.py (hunk 3 above)
Comment (HermitSun) [reply python/paddle/distributed/collective.py:983]: Revised.

**Review by LiYuRio — DISMISSED** [review]

**Comment by HermitSun** [comment]
> To add support of `bool` type, should we modify c++ backend?

Maybe we have supported `bool` type in prev pr https://github.com/PaddlePaddle/Paddle/pull/44417.

**Review by liuTINA0907 — DISMISSED** [review]

**Review by liuTINA0907 — DISMISSED** [review]

**Review by LiYuRio — APPROVED** [review]

**Review by liuTINA0907 — APPROVED** [review]

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**